command-line/lib/Command/Script.hs

```diff
@@ -0,0 +1,134 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+{-# OPTIONS_HADDOCK hide #-}
+
+module Command.Script
+    ( Cmd
+    , mod
+    , run
+
+    -- * Internal
+    , requireSignatureOfParser
+    , requireAllOfParser
+    , requireAnyOfParser
+    , requireAtLeastOfParser
+    , scriptParser
+    ) where
+
+import Cardano.Multisig
+    ( Script (..), toScriptHash, verificationKeyHashFromBytes )
+import Codec.Binary.Bech32.TH
+    ( humanReadablePart )
+import Codec.Binary.Encoding
+    ( fromBase16 )
+import Data.Char
+    ( isDigit, isLetter )
+import Data.Word
+    ( Word8 )
+import Options.Applicative
+    ( CommandFields, Mod, command, footerDoc, header, helper, info, progDesc )
+import Options.Applicative.Encoding
+    ( Encoding, encodingOpt )
+import Options.Applicative.Help.Pretty
+    ( bold, indent, string, vsep )
+import Prelude hiding
+    ( mod )
+import System.IO
+    ( stdin, stdout )
+import System.IO.Extra
+    ( hPutBytes, noNewline, progName )
+import Text.ParserCombinators.ReadP
+    ( ReadP, readP_to_S, (<++) )
+
+import qualified Data.ByteString.Char8 as B8
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import qualified Text.ParserCombinators.ReadP as P
+
+
+newtype Cmd = Cmd
+    { encoding :: Encoding
+    } deriving (Show)
+
+mod :: (Cmd -> parent) -> Mod CommandFields parent
+mod liftCmd = command "scripthash" $
+    info (helper <*> fmap liftCmd parser) $ mempty
+        <> progDesc "Create a script hash"
+        <> header "Create a script hash \
+            \that can be used in stake or payment address."
+        <> footerDoc (Just $ vsep
+            [ string "The script is read from stdin."
+            , string ""
+            , string "Example:"
+            , indent 2 $ bold $ string "$ cat script.txt \\"
+            , indent 4 $ bold $ string $ "| "<>progName<>" scripthash \\"
+            , indent 2 $ string "3c07030e36bfffe67e2e2ec09e5293d384637cd2f004356ef320f3fe"
+            ])
+  where
+    parser = Cmd
+        <$> encodingOpt [humanReadablePart|xpub|]
+
+run :: Cmd -> IO ()
+run Cmd{encoding} = do
+    bytes <- B8.filter noNewline <$> B8.hGetContents stdin
+    case readP_to_S scriptParser (B8.unpack bytes) of
+         [(multisig,_rest)] ->
+             hPutBytes stdout (toScriptHash multisig) encoding
+         _ ->
+             error "parsing the script failed"
+
+scriptParser :: ReadP Script
```

Comment:
Is this script syntax documented anywhere?

The parser should go in a separate module.

Also I don't think there would be any problem with using a more featureful parsing library than ReadP, if you want.

Reply:
added descr here https://github.com/input-output-hk/cardano-addresses/pull/76/commits/8b630990fe95d9972edecd2f26db0494c46fc45f
when it comes to ReadP the advantage of it is that it is already in base, no need for any additional dependency. And I think it serves well in this particular context (side note: it was positive surprise when discovering it)

Reply:
I was wondering, could you just use the automatically derived Read and Show instances for scripts?

Reply:
I think that both the Json syntax we'll use in Cardano-wallet and this string dsl should be speficied in the CIP.